openmsupply/application-manager-server
# #439 Carbone template for Feature Showcase

Fixes #439 

Having done this, I would say it's definitely too complicated to try and make a "general" carbone template for all templates. Every question is different -- not just different question types, but different instances of the same question type require case-specific formatting.

Here is the carbone template for Feature Showcase: [feature_showcase.odt](https://github.com/openmsupply/application-manager-server/files/6875752/feature_showcase.odt)

Here is an example output PDF: [feature_showcase_S-VKT-0100_UlUggoa7IIjW0cz9urfrN.pdf](https://github.com/openmsupply/application-manager-server/files/6875809/feature_showcase_S-VKT-0100_UlUggoa7IIjW0cz9urfrN.pdf)


### How to test

Easiest way is to use Postman to manually run the Actions. Here is a collection with the endpoints and parameters you'll need: 
[Application Manager.postman_collection.json.zip](https://github.com/openmsupply/application-manager-server/files/6875759/Application.Manager.postman_collection.json.zip)

1. Start a new Feature Showcase application. You don't need to submit it, just make sure responses have been added (it's probably best not to submit, so you can quickly change responses and see how the PDF output changes)
2. Use the `/upload` endpoint to upload the carbone template file and return a uniqueId
3. Use the `/run-action` endpoint  to "generateDoc", using your application ID and the previous uniqueId (for the file), like so:
```js
{
    "actionCode": "generateDoc",
    "applicationId": 1,
    "parameters": { 
        "docTemplateId": "vXQzU3H3EcFOOSmgiDady"
    }
}
```
4. Open the created PDF and observe the results :) 

### Additional changes

- added an "additionalData" parameter to "generateDoc" (similar to "generateTextString"), in case we want to pass data other than "applicationData" and "outputCumulative"
- "applicationData" now contains "orgName" as well (acquired as part of `getUserData`)
- fix for "subfolder" parameter in `/upload` endpoint (can use either "subfolder" or "subFolder", or "sub_folder")
